File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/MaximumSegments.c
#include "java_arrays.h"
#include <stdint.h>

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left - right);
*/
static int32_t java_sub(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left - (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (left > right ? left : right);
*/
static int32_t java_max(int32_t left, int32_t right) {
    return left > right ? left : right;
}

/*@
  logic integer fb_extend(integer value) = value < 0 ? -1 : value + 1;
  logic integer fb_segments(integer n, integer a, integer b, integer c) =
    n < 0 ? -1 : n == 0 ? 0 :
    \max(fb_extend(fb_segments(n - a, a, b, c)),
         \max(fb_extend(fb_segments(n - b, a, b, c)), fb_extend(fb_segments(n - c, a, b, c))));
*/
/*@
  requires 0 <= n < INT32_MAX && a > 0 && b > 0 && c > 0;
  assigns \nothing;
  ensures \result == fb_segments(n, a, b, c);
  ensures -1 <= \result <= n;
*/
int32_t maximumSegments(int32_t n, int32_t a, int32_t b, int32_t c) {
    JIntArray dp = jarray_new(java_add(n, 1));
    jarray_set(dp, 0, 0);
    /*@
      loop invariant 1 <= i <= (integer)n + 1;
      loop invariant jintarray_valid_nonnull(dp) && dp->length == n + 1;
      loop invariant \forall integer k; 0 <= k < i ==> dp->data[k] == fb_segments(k, a, b, c) && -1 <= dp->data[k] <= k;
      loop assigns i, dp->data[0 .. n];
      loop variant (integer)n - i + 1;
    */
    for (int32_t i = 1; i <= n; i = java_add(i, 1)) {
        jarray_set(dp, i, -1);
        if (i >= a && jarray_get(dp, java_sub(i, a)) != -1) {
            jarray_set(dp, i, java_max(jarray_get(dp, i),
                                       java_add(jarray_get(dp, java_sub(i, a)), 1)));
        }
        if (i >= b && jarray_get(dp, java_sub(i, b)) != -1) {
            jarray_set(dp, i, java_max(jarray_get(dp, i),
                                       java_add(jarray_get(dp, java_sub(i, b)), 1)));
        }
        if (i >= c && jarray_get(dp, java_sub(i, c)) != -1) {
            jarray_set(dp, i, java_max(jarray_get(dp, i),
                                       java_add(jarray_get(dp, java_sub(i, c)), 1)));
        }
    }
    return jarray_get(dp, n);
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/MaximumSegments.java
import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;


class MaximumSegments {
    
    /*@ public normal_behavior
      @ ensures \result == (a >= b ? a : b);
      @ model public static pure \bigint fb_max(\bigint a, \bigint b);
      @*/
    /*@ public normal_behavior
      @ requires value >= -1;
      @ ensures \result == (value == -1 ? -1 : value + 1);
      @ model public static pure \bigint fb_extend(\bigint value);
      @*/
    /*@ public normal_behavior
      @ requires a > 0 && b > 0 && c > 0;
      @ ensures \result == (n < 0 ? -1 : n == 0 ? 0 : fb_max(fb_extend(fb_segments(n - a, a, b, c)), fb_max(fb_extend(fb_segments(n - b, a, b, c)), fb_extend(fb_segments(n - c, a, b, c)))));
      @ ensures \result >= -1;
      @ measured_by (n > 0 ? n : 0);
      @ model public static pure \bigint fb_segments(\bigint n, int a, int b, int c);
      @*/
    /*@ public normal_behavior
      @ requires 0 <= n < Integer.MAX_VALUE && a > 0 && b > 0 && c > 0;
      @ assignable \nothing;
      @ ensures \result == fb_segments(n, a, b, c);
      @ ensures -1 <= \result <= n;
      @*/
    //@ code_java_math
    public static int maximumSegments(int n, int a, int b, int c) {
        int dp[] = new int[n+1];
        dp[0] = 0;
        /*@ loop_invariant 1 <= i <= n + 1;
          @ loop_invariant (\forall int k; 0 <= k < i; dp[k] == fb_segments(k, a, b, c) && -1 <= dp[k] <= k);
          @ loop_writes i, dp[*];
          @ decreases (\bigint)n - i + 1;
          @*/
        for(int i = 1; i <= n; i++){
            dp[i] = -1;
            if(i >= a && dp[i-a] != -1){
                dp[i] = Math.max(dp[i], dp[i-a]+1);
            }
            if(i >= b && dp[i-b] != -1){
                dp[i] = Math.max(dp[i], dp[i-b]+1);
            }
            if(i >= c && dp[i-c] != -1){
                dp[i] = Math.max(dp[i], dp[i-c]+1);
            }
        }
        return dp[n];
    }
}

